zipkin-autoconfigure/storage-cassandra3/src/main/java/zipkin/autoconfigure/storage/cassandra3/brave/TracedSession.java

```diff
@@ -11,204 +11,182 @@
-      }
-      return;
+
+    @Override
+    public String getLoggedKeyspace() {
+        return session.getLoggedKeyspace();
     }
-    span.setDuration(nanos / 1000); // TODO: allow client tracer to end with duration
-    Endpoint local = span.getAnnotations().get(0).host; // TODO: expose in brave
-    long endTs = span.getTimestamp() + span.getDuration();
-    span.addToAnnotations(Annotation.create(endTs, "cr", local));
-    if (e != null) {
-      span.addToBinary_annotations(BinaryAnnotation.create(Constants.ERROR, e.getMessage(), local));
+
+    @Override
+    public Session init() {
+        return session.init();
     }
-    int ipv4 = ByteBuffer.wrap(host.getAddress().getAddress()).getInt();
-    Endpoint endpoint = Endpoint.create("cassandra3", ipv4, host.getSocketAddress().getPort());
-    span.addToBinary_annotations(BinaryAnnotation.address("sa", endpoint));
-    collector.collect(span);
-  }
-
-  @Override
-  public boolean equals(Object obj) {
-    if (obj instanceof TracedSession) {
-      TracedSession other = (TracedSession) obj;
-      return target.equals(other.target);
+
+    @Override
+    public ResultSetFuture executeAsync(final Statement statement) {
+        // don't trace any already self-traced spans
+        boolean alreadySelfTraced = null != statement.getOutgoingPayload()
+                && statement.getOutgoingPayload().containsKey(Cassandra3Storage.SELF_TRACING_KEY);
+
+        // o.a.c.tracing.Tracing.newSession must use the same format for the key zipkin
+        boolean supported = version.compareTo(ProtocolVersion.V4) >= 0;
+        if (supported && !alreadySelfTraced) {
+            useKeyspace(session, statement.getKeyspace());
+            SpanId spanId = null;
+            spanId = clientTracer.startNewSpan(statement.toString());
+            clientTracer.submitBinaryAnnotation(Cassandra3Storage.SELF_TRACING_KEY, "true");
+            if (null != spanId) {
+                String ks = null != statement.getKeyspace() ? statement.getKeyspace() : session.getLoggedKeyspace();
+                assert null != ks : "this cluster/session only for keyspace " + ks;
+                clientTracer.setClientSent();
+            }
+            statement.enableTracing();
+            Map<String,ByteBuffer> payload = new HashMap<>();
+            if (null != statement.getOutgoingPayload()) {
+                payload.putAll(statement.getOutgoingPayload());
+            }
+            payload.put("zipkin", ByteBuffer.wrap(spanId.bytes()));
+            statement.setOutgoingPayload(payload);
+
+            final Span span = null != spanId ? brave.clientSpanThreadBinder().getCurrentClientSpan() : null;
+            ResultSetFuture future = session.executeAsync(statement);
+            if (null != spanId) {
+                future.addListener(new Runnable() {
+                    @Override
+                    public void run() {
+                        brave.clientSpanThreadBinder().setCurrentSpan(span);
```

Comment:
albeit unlikely the middle bits will throw, I'd add a try-finally. Also note server-modules are language level 8, so you can do a lambda